### PR TITLE
Add Guard Clauses with Parameter Arguments (Issue #155)

### PR DESCRIPTION
**Pull request by trgrote**
Add support for guard clauses as those found in the *If() configurations to support arguments passed in via a TriggerWithParameters trigger.  Similar to how the Dynamic functions handled passed in arguments. 

Example:
```csharp
var sm = new StateMachine<State, Trigger>(State.A);
var x = sm.SetTriggerParameters<int>(Trigger.X);
sm.Configure(State.A).PermitIf(x, State.B, i => i == 2);
sm.Fire(x, 1);    // Does not transition
sm.Fire(x, 2);    // transitions to state B
```

Add the following functions to StateConfiguration:
```csharp
InternalTransitionIf<TArg0>(TriggerWithParameters<TArg0> trigger, Func<TArg0, bool> guard, Action<TArg0, Transition> internalAction)
InternalTransitionIf<TArg0, TArg1>(TriggerWithParameters<TArg0, TArg1> trigger, Func<TArg0, TArg1, bool> guard, Action<TArg0, TArg1, Transition> internalAction)
InternalTransitionIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, Func<TArg0, TArg1, TArg2, bool> guard, Action<TArg0, TArg1, TArg2, Transition> internalAction)

PermitIf<TArg0>(TriggerWithParameters<TArg0> trigger, TState destinationState, Func<TArg0, bool> guard, string guardDescription = null)
PermitIf<TArg0>(TriggerWithParameters<TArg0> trigger, TState destinationState, params Tuple<Func<TArg0, bool>, string>[] guards)
PermitIf<TArg0, TArg1>(TriggerWithParameters<TArg0, TArg1> trigger, TState destinationState, Func<TArg0, TArg1, bool> guard, string guardDescription = null)
PermitIf<TArg0, TArg1>(TriggerWithParameters<TArg0, TArg1> trigger, TState destinationState, params Tuple<Func<TArg0, TArg1, bool>, string>[] guards)
PermitIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, TState destinationState, Func<TArg0, TArg1, TArg2, bool> guard, string guardDescription = null)
PermitIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, TState destinationState, params Tuple<Func<TArg0, TArg1, TArg2, bool>, string>[] guards)

PermitReentryIf<TArg0>(TriggerWithParameters<TArg0> trigger, Func<TArg0, bool> guard, string guardDescription = null)
PermitReentryIf<TArg0>(TriggerWithParameters<TArg0> trigger, params Tuple<Func<TArg0, bool>, string>[] guards)
PermitReentryIf<TArg0, TArg1>(TriggerWithParameters<TArg0, TArg1> trigger, Func<TArg0, TArg1, bool> guard, string guardDescription = null)
PermitReentryIf<TArg0, TArg1>(TriggerWithParameters<TArg0, TArg1> trigger, params Tuple<Func<TArg0, TArg1, bool>, string>[] guards)
PermitReentryIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, Func<TArg0, TArg1, TArg2, bool> guard, string guardDescription = null)
PermitReentryIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, params Tuple<Func<TArg0, TArg1, TArg2, bool>, string>[] guards)

IgnoreIf<TArg0>(TriggerWithParameters<TArg0> trigger, Func<TArg0, bool> guard, string guardDescription = null)
IgnoreIf<TArg0>(TriggerWithParameters<TArg0> trigger, params Tuple<Func<TArg0, bool>, string>[] guards)
IgnoreIf<TArg0, TArgo1>(TriggerWithParameters<TArg0, TArgo1> trigger, Func<TArg0, TArgo1, bool> guard, string guardDescription = null)
IgnoreIf<TArg0, TArg1>(TriggerWithParameters<TArg0, TArg1> trigger, params Tuple<Func<TArg0, TArg1, bool>, string>[] guards)
IgnoreIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, Func<TArg0, TArg1, TArg2, bool> guard, string guardDescription = null)
IgnoreIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, params Tuple<Func<TArg0, TArg1, TArg2, bool>, string>[] guards)

PermitDynamicIf<TArg0>(TriggerWithParameters<TArg0> trigger, Func<TArg0, TState> destinationStateSelector, Func<TArg0, bool> guard, string guardDescription = null)
PermitDynamicIf<TArg0>(TriggerWithParameters<TArg0> trigger, Func<TArg0, TState> destinationStateSelector, params Tuple<Func<TArg0, bool>, string>[] guards)
PermitDynamicIf<TArg0, TArg1>(TriggerWithParameters<TArg0, TArg1> trigger, Func<TArg0, TArg1, TState> destinationStateSelector, Func<TArg0, TArg1, bool> guard, string guardDescription = null)
PermitDynamicIf<TArg0, TArg1>(TriggerWithParameters<TArg0, TArg1> trigger, Func<TArg0, TArg1, TState> destinationStateSelector, Tuple<Func<TArg0, TArg1, bool>, string>[] guards)
PermitDynamicIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, Func<TArg0, TArg1, TArg2, TState> destinationStateSelector, Func<TArg0, TArg1, TArg2, bool> guard, string guardDescription = null)
PermitDynamicIf<TArg0, TArg1, TArg2>(TriggerWithParameters<TArg0, TArg1, TArg2> trigger, Func<TArg0, TArg1, TArg2, TState> destinationStateSelector, Tuple<Func<TArg0, TArg1, TArg2, bool>, string>[] guards)

```

I needed to make changes to how guard clauses function in order to support optional arguments.  Because of that, the underlying guard clause type is now ```Func<object[], bool>``` instead of ```Func<bool>```.  Old guard clauses are still compatible, they just have a 0 arg object array passed in.  This change is backwards compatible.

### CODE REVIEW

File: src/Stateless/StateRepresentation.cs [hunk 1]
```diff
@@ -45,18 +45,18 @@ internal ICollection<StateRepresentation> GetSubstates()
...
-                    (Superstate != null && Superstate.TryFindHandler(trigger, out handler)));
+                return (TryFindLocalHandler(trigger, args, out handler) ||
+                    (Superstate != null && Superstate.TryFindHandler(trigger, args, out handler)));
             }
 
-            bool TryFindLocalHandler(TTrigger trigger, out TriggerBehaviourResult handlerResult)
+            bool TryFindLocalHandler(TTrigger trigger, object[] args, out TriggerBehaviourResult handlerResult)
             {
                 if (!_triggerBehaviours.TryGetValue(trigger, out ICollection<TriggerBehaviour> possible))
                 {
```
Comment (HenningNT) [inline src/Stateless/StateRepresentation.cs]: I get a compiler error on line 69, (h.UnmetGuardConditions):
Error	CS1503	Argument 2: cannot convert from 'method group' to 'ICollection<string>'	Stateless(net40), Stateless(net45), Stateless(netstandard1.0)	C:\Users\Henning.Torsteinsen\Source\Repos\trgrote-stateless\src\Stateless\StateRepresentation.cs	Active	69

**Comment by HenningNT** [comment]
I'm afraid it doesn't build for me...

**Comment by trgrote** [comment]
Ok, I'll look at it

File: src/Stateless/StateRepresentation.cs [hunk 2]
```diff
@@ -66,7 +66,7 @@ bool TryFindLocalHandler(TTrigger trigger, object[] args, out TriggerBehaviourRe
 
                 // Guard functions executed
                 var actual = possible
-                    .Select(h => new TriggerBehaviourResult(h, h.UnmetGuardConditions(args)));
```
Comment (trgrote) [inline src/Stateless/StateRepresentation.cs]: Ah this merge screwed it up.

**Comment by HenningNT** [comment]
There's a lot to go through here. I'll check with one of my projects.
Does this change affect the DotGraph output?

**Comment by igor-eremyashev** [comment]
Very useful and much anticipated feature, hope it will be merged to the base branch soon.

**Comment by HenningNT** [comment]
@igor-eremyashev would you mind giving it a test run?

**Comment by igor-eremyashev** [comment]
@HenningNT 
I created a unit test to test this feature, worked as expected for me. I attached the code I used to test PermitIf with parameters.
https://gist.github.com/igor-eremyashev/4af56d0c3abb3eea0264a3804660b179

**Comment by HenningNT** [comment]
ok, thanks. I'll take a chance and merge this now.

**Comment by igor-eremyashev** [comment]
@HenningNT Thank you for the quick merge! I see that AppVeyor build is failing because of the following error:
"Specify which project or solution file to use because the folder contains more than one project or solution file."
I think this is because there are 2 .sln files in the repository - one in root folder and another one in Backup folder.
Could you try to change the extension of the solution file in Backup folder (or remove it if possible)?
Would be great to be able to install the latest version from Nuget.

**Comment by HenningNT** [comment]
Backup folder deleted.
... hope I didn't cause any new problems :-)

**Comment by HenningNT** [comment]
Looks like something else is broken now :-(

**Comment by igor-eremyashev** [comment]
Probably something in appveyor configuration is broken, but I don't know what it could be.
For example, in the latest successful build
https://ci.appveyor.com/project/DotnetStateMachine/stateless/build/261
there is output from Build.ps1 (like 'build: Build started', dotnet cli bootstrapping) and in there current build 
https://ci.appveyor.com/project/DotnetStateMachine/stateless
there nothing on it. Like like build.ps1 haven't been called at all.

**Comment by dmitrynovik** [comment]
Any indication when this PR will be merged into the master branch so that we could consume it with Nuget? Looks like master is quite stale: last commit has been made on Sepember 2017.

Cheers.

**Comment by HenningNT** [comment]
The AppVeyor build is failing, and I have no idea how to fix it. Maybe @nblumhardt could take a look?

**Comment by nblumhardt** [comment]
Hi @HenningNT! Hope all's well with you :wave: 

Just had a really quick skim, it looks like the appveyor.yml install script is still installing the .NET Core 1.0 SDK: https://github.com/dotnet-state-machine/stateless/blob/dev/appveyor.yml#L7

I think removing the install section altogether should be possible, here.

Not sure it will help, but happy to keep digging in if that doesn't help :+1:

**Comment by HenningNT** [comment]
Things are good, I'm starting working for [Bouvet](www.bouvet.no) in a couple of weeks.

I'm struggling a bit with this build stuff... I tried removing everything from the install section, but that didn't help... Is the build.ps1 script necessary?

**Comment by nblumhardt** [comment]
As far as I can tell, AppVeyor's using the MSBuild 14 engine rather than 15:

```
Microsoft (R) Build Engine version 14.0.25420.1
```

Trying to remember how we force it to use 15....

**Comment by nblumhardt** [comment]
Found it - the GitHub authorization had failed, so it wasn't picking up appveyor.yml.

I authorized as me, and ran a dev build, then revoked; we need to set up a "build" user for this org. Can help when I have a chance :-)

**Comment by HenningNT** [comment]
Please do!

**Pull request merged**